### PR TITLE
Instrumentation api update

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: src/main/java/org/truffleruby/language/RubyNode.java [hunk 1]
```diff
@@ -36,8 +42,38 @@ public Object isDefined(VirtualFrame frame) {
...
+        if (tag == TraceManager.LineTag.class) {
+            return isNewLine() || isCoverageLine();
+        }
+
+        if (tag == StandardTags.RootTag.class) {
+            return isRoot();
+        }
+
+        return false;
+    }
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/RubyNode.java]: Is it intentional to move hasTag() here instead of in RubyBaseNode?

File: src/main/java/org/truffleruby/language/RubyNode.java [hunk 2]
```diff
@@ -36,8 +42,38 @@ public Object isDefined(VirtualFrame frame) {
...
+        if (tag == TraceManager.CallTag.class || tag == StandardTags.CallTag.class) {
+            return isCall();
+        }
+
+        if (tag == StandardTags.StatementTag.class) {
+            return isNewLine();
+        }
+
+        if (tag == TraceManager.LineTag.class) {
+            return isNewLine() || isCoverageLine();
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/RubyNode.java]: Why is this changed?

File: src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java [hunk 3]
```diff
@@ -11,77 +11,64 @@
...
+                break;
             } catch (Throwable t) {
-                probeNode.onReturnExceptional(frame, t);
+                Object result = probeNode.onReturnExceptionalOrUnwind(frame, t, wasOnReturnExecuted);
+                if (result == ProbeNode.UNWIND_ACTION_REENTER) {
+                    continue;
+                } else if (result != null) {
+                    returnValue = result;
+                    break;
+                }
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java]: Is this a documented pattern or that same as the generated wrapper?
Looks like `return` could be used in the loop directly instead of `returnValue` (then it can a local variable only in the `try` part).

File: src/main/java/org/truffleruby/core/CoreLibrary.java [hunk 4]
```diff
@@ -525,7 +529,11 @@ public CoreLibrary(RubyContext context) {
         defineModule(truffleModule, "ReadlineHistory");
         defineModule(truffleModule, "ThreadOperations");
         psychModule = defineModule("Psych");
-        psychParserClass = defineClass(psychModule, objectClass, "Parser");
+        publicPsychParserClass = defineClass(psychModule, objectClass, "Parser");
+        psychYAMLParserClass = defineClass(psychModule, objectClass, "YAMLParser");
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/CoreLibrary.java]: I think those should be defined under `Psych::Parser`, not under `Psych` directly.

**Comment by chrisseaton** [comment]
Is `NotProvided` a problem as well? That isn't a `TruffleObject`.

File: src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java [hunk 5]
```diff
@@ -166,28 +165,38 @@ private StreamReader newStreamReader(DynamicObject io) {
     @CoreMethod(names = "get_event", needsSelf = false, required = 1)
     public abstract static class GetEventNode extends CoreMethodArrayArgumentsNode {
         @TruffleBoundary
-        @Specialization
-        protected Event getEvent(Parser parser) {
-            return parser.getEvent();
+        @Specialization(guards = "isParser(parser)")
+        protected DynamicObject getEvent(DynamicObject parser) {
```
Comment (wirthi) [inline src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java]: I guess the usage of this node guarantees that the guard always matches?

Comment (eregon) [reply src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java]: Yes, but I think it's still good to have the guard to prevent accidental misuses.

File: src/main/java/org/truffleruby/language/RubyNode.java (hunk 1 above)
Comment (aardvark179) [reply src/main/java/org/truffleruby/language/RubyNode.java]: Yes. `RubyBaseNode` previously overrode `isTaggedWith()` on `Node` but did not mark itself as instrumentable. `hasTag()` is now defined on `InstrumentableNode`, and since `RubyNode` is the most basic node type we do mark as instrumentable that is the class I chose to implement `InstrumentableNode` on, and so it made sense to move the method to there.

File: src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java (hunk 3 above)
Comment (aardvark179) [reply src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java]: It's the same as the generated code. It would be nice if we didn't have to manually generate the wrapper here, but we need to for `canHandle()`.

File: src/main/java/org/truffleruby/core/CoreLibrary.java (hunk 4 above)
Comment (aardvark179) [reply src/main/java/org/truffleruby/core/CoreLibrary.java]: I'll move them.

**Comment by aardvark179** [comment]
@chrisseaton `NotProvided` shouldn't be an issue as it's not returned by anything.

File: src/main/java/org/truffleruby/language/RubyNode.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/RubyNode.java]: Makes sense.

File: src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java (hunk 3 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java]: Can you add a comment about why this wrapper is needed at the top of the class?

**Comment by eregon** [comment]
@aardvark179 `undefined` is the same as `NotProvided.INSTANCE` in Ruby code. But maybe it's never returned from a method or an instrumentable node. If the gate passes, it should be fine.

File: src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java (hunk 5 above)
Comment (wirthi) [reply src/main/java/org/truffleruby/stdlib/psych/PsychParserNodes.java]: Sure, if it is clear error, than it's fine.
